File: compiler/packages/babel-plugin-react-forget/src/Optimization/ConstantPropagation.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { isValidIdentifier } from "@babel/types";
import { CompilerError } from "../CompilerError";
import {
  Environment,
  GotoVariant,
  HIRFunction,
  IdentifierId,
  Instruction,
  InstructionValue,
  LoadGlobal,
  Phi,
  Place,
  Primitive,
  assertConsistentIdentifiers,
  assertTerminalSuccessorsExist,
  markInstructionIds,
  markPredecessors,
  mergeConsecutiveBlocks,
  removeUnreachableFallthroughs,
  reversePostorderBlocks,
} from "../HIR";
import {
  removeDeadDoWhileStatements,
  removeUnnecessaryTryCatch,
  removeUnreachableForUpdates,
} from "../HIR/HIRBuilder";
import { eliminateRedundantPhi } from "../SSA";

/**
 * Applies constant propagation and constant folding to the given function.
 * Note that because HIR operands are always a Place, constants cannot be directly
 * propagated into the HIR itself (the closest option would be to copy constants to
 * new temporaries just before each use, and update usage sites to reference those
 * new temporaries).
 *
 * Instead this pass implements constant folding, in which constant values are
 * propagated internally to the pass and subsequent operations are removed/folded where
 * possible.
 *
 * Note that this pass may prune control flow blocks that are unreachable, for example
 * a consequent or alternate branch if an `if` test is provably truthy or falsey.
 * If (and only if) terminals change, the pass re-runs various stages to ensure the
 * CFG is in minimal form. This means instruction ids *may* change as a result of this
 * pass.
 */
export function constantPropagation(fn: HIRFunction): void {
  const constants: Constants = new Map();
  constantPropagationImpl(fn, constants);
}

function constantPropagationImpl(fn: HIRFunction, constants: Constants): void {
  while (true) {
    const haveTerminalsChanged = applyConstantPropagation(fn, constants);
    if (!haveTerminalsChanged) {
      break;
    }
    // If terminals have changed then blocks may have become newly unreachable.
    // Re-run minification of the graph (incl reordering instruction ids)
    reversePostorderBlocks(fn.body);
    removeUnreachableFallthroughs(fn.body);
    removeUnreachableForUpdates(fn.body);
    removeDeadDoWhileStatements(fn.body);
    removeUnnecessaryTryCatch(fn.body);
    markInstructionIds(fn.body);
    markPredecessors(fn.body);

    // Now that predecessors are updated, prune phi operands that can never be reached
    for (const [, block] of fn.body.blocks) {
      for (const phi of block.phis) {
        for (const [predecessor] of phi.operands) {
          if (!block.preds.has(predecessor)) {
            phi.operands.delete(predecessor);
          }
        }
      }
    }
    // By removing some phi operands, there may be phis that were not previously
    // redundant but now are
    eliminateRedundantPhi(fn);
    // Finally, merge together any blocks that are now guaranteed to execute
    // consecutively
    mergeConsecutiveBlocks(fn);

    assertConsistentIdentifiers(fn);
    assertTerminalSuccessorsExist(fn);
  }
}

function applyConstantPropagation(
  fn: HIRFunction,
  constants: Constants
): boolean {
  let hasChanges = false;
  for (const [, block] of fn.body.blocks) {
    // Initialize phi values if all operands have the same known constant value.
    // Note that this analysis uses a single-pass only, so it will never fill in
    // phi values for blocks that have a back-edge.
    for (const phi of block.phis) {
      let value = evaluatePhi(phi, constants);
      if (value !== null) {
        constants.set(phi.id.id, value);
      }
    }

    for (let i = 0; i < block.instructions.length; i++) {
      if (block.kind === "sequence" && i === block.instructions.length - 1) {
        // evaluating the last value of a value block can break order of evaluation,
        // skip these instructions
        continue;
      }
      const instr = block.instructions[i]!;
      const value = evaluateInstruction(fn.env, constants, instr);
      if (value !== null) {
        constants.set(instr.lvalue.identifier.id, value);
      }
    }

    const terminal = block.terminal;
    switch (terminal.kind) {
      case "if": {
        const testValue = read(constants, terminal.test);
        if (testValue !== null && testValue.kind === "Primitive") {
          hasChanges = true;
          const targetBlockId = testValue.value
            ? terminal.consequent
            : terminal.alternate;
          block.terminal = {
            kind: "goto",
            variant: GotoVariant.Break,
            block: targetBlockId,
            id: terminal.id,
            loc: terminal.loc,
          };
        }
        break;
      }
      default: {
        // no-op
      }
    }
  }

  return hasChanges;
}

function evaluatePhi(phi: Phi, constants: Constants): Constant | null {
  let value: Constant | null = null;
  for (const [, operand] of phi.operands) {
    const operandValue = constants.get(operand.id) ?? null;
    // did not find a constant, can't constant propogate
    if (operandValue === null) {
      return null;
    }

    // first iteration of the loop, let's store the operand and continue
    // looping.
    if (value === null) {
      value = operandValue;
      continue;
    }

    // found different kinds of constants, can't constant propogate
    if (operandValue.kind !== value.kind) {
      return null;
    }

    switch (operandValue.kind) {
      case "Primitive": {
        CompilerError.invariant(value.kind === "Primitive", {
          reason: "value kind expected to be Primitive",
          loc: null,
          suggestions: null,
        });

        // different constant values, can't constant propogate
        if (operandValue.value !== value.value) {
          return null;
        }
        break;
      }
      case "LoadGlobal": {
        CompilerError.invariant(value.kind === "LoadGlobal", {
          reason: "value kind expected to be LoadGlobal",
          loc: null,
          suggestions: null,
        });

        // different global values, can't constant propogate
        if (operandValue.name !== value.name) {
          return null;
        }
        break;
      }
      default:
        return null;
    }
  }

  return value;
}

function evaluateInstruction(
  env: Environment,
  constants: Constants,
  instr: Instruction
): Constant | null {
  const value = instr.value;
  switch (value.kind) {
    case "Primitive": {
      return value;
    }
    case "LoadGlobal": {
      return value;
    }
    case "ComputedLoad": {
      const property = read(constants, value.property);
      if (
        property !== null &&
        property.kind === "Primitive" &&
        typeof property.value === "string" &&
        isValidIdentifier(property.value)
      ) {
        const nextValue: InstructionValue = {
          kind: "PropertyLoad",
          loc: value.loc,
          property: property.value,
          object: value.object,
        };
        instr.value = nextValue;
      }
      return null;
    }
    case "ComputedStore": {
      const property = read(constants, value.property);
      if (
        property !== null &&
        property.kind === "Primitive" &&
        typeof property.value === "string" &&
        isValidIdentifier(property.value)
      ) {
        const nextValue: InstructionValue = {
          kind: "PropertyStore",
          loc: value.loc,
          property: property.value,
          object: value.object,
          value: value.value,
        };
        instr.value = nextValue;
      }
      return null;
    }
    case "PostfixUpdate": {
      const previous = read(constants, value.value);
      if (
        previous !== null &&
        previous.kind === "Primitive" &&
        typeof previous.value === "number"
      ) {
        const next =
          value.operation === "++" ? previous.value + 1 : previous.value - 1;
        // Store the updated value
        constants.set(value.lvalue.identifier.id, {
          kind: "Primitive",
          value: next,
          loc: value.loc,
        });
        // But return the value prior to the update
        return previous;
      }
      return null;
    }
    case "PrefixUpdate": {
      const previous = read(constants, value.value);
      if (
        previous !== null &&
        previous.kind === "Primitive" &&
        typeof previous.value === "number"
      ) {
        const next: Primitive = {
          kind: "Primitive",
          value:
            value.operation === "++" ? previous.value + 1 : previous.value - 1,
          loc: value.loc,
        };
        // Store and return the updated value
        constants.set(value.lvalue.identifier.id, next);
        return next;
      }
      return null;
    }
    case "BinaryExpression": {
      const lhsValue = read(constants, value.left);
      const rhsValue = read(constants, value.right);
      if (
        lhsValue !== null &&
        rhsValue !== null &&
        lhsValue.kind === "Primitive" &&
        rhsValue.kind === "Primitive"
      ) {
        const lhs = lhsValue.value;
        const rhs = rhsValue.value;
        let result: Primitive | null = null;
        switch (value.operator) {
          case "+": {
            if (typeof lhs === "number" && typeof rhs === "number") {
              result = { kind: "Primitive", value: lhs + rhs, loc: value.loc };
            }
            break;
          }
          case "-": {
            if (typeof lhs === "number" && typeof rhs === "number") {
              result = { kind: "Primitive", value: lhs - rhs, loc: value.loc };
            }
            break;
          }
          case "*": {
            if (typeof lhs === "number" && typeof rhs === "number") {
              result = { kind: "Primitive", value: lhs * rhs, loc: value.loc };
            }
            break;
          }
          case "/": {
            if (typeof lhs === "number" && typeof rhs === "number") {
              result = { kind: "Primitive", value: lhs / rhs, loc: value.loc };
            }
            break;
          }
          case "<": {
            if (typeof lhs === "number" && typeof rhs === "number") {
              result = { kind: "Primitive", value: lhs < rhs, loc: value.loc };
            }
            break;
          }
          case "<=": {
            if (typeof lhs === "number" && typeof rhs === "number") {
              result = { kind: "Primitive", value: lhs <= rhs, loc: value.loc };
            }
            break;
          }
          case ">": {
            if (typeof lhs === "number" && typeof rhs === "number") {
              result = { kind: "Primitive", value: lhs > rhs, loc: value.loc };
            }
            break;
          }
          case ">=": {
            if (typeof lhs === "number" && typeof rhs === "number") {
              result = { kind: "Primitive", value: lhs >= rhs, loc: value.loc };
            }
            break;
          }
          case "==": {
            result = { kind: "Primitive", value: lhs == rhs, loc: value.loc };
            break;
          }
          case "===": {
            result = { kind: "Primitive", value: lhs === rhs, loc: value.loc };
            break;
          }
          case "!=": {
            result = { kind: "Primitive", value: lhs != rhs, loc: value.loc };
            break;
          }
          case "!==": {
            result = { kind: "Primitive", value: lhs !== rhs, loc: value.loc };
            break;
          }
          default: {
            break;
          }
        }
        if (result !== null) {
          instr.value = result;
          return result;
        }
      }
      return null;
    }
    case "PropertyLoad": {
      const objectValue = read(constants, value.object);
      if (objectValue !== null) {
        if (
          objectValue.kind === "Primitive" &&
          typeof objectValue.value === "string" &&
          value.property === "length"
        ) {
          const result: InstructionValue = {
            kind: "Primitive",
            value: objectValue.value.length,
            loc: value.loc,
          };
          instr.value = result;
          return result;
        }
      }
      return null;
    }
    case "LoadLocal": {
      const placeValue = read(constants, value.place);
      if (placeValue !== null) {
        instr.value = placeValue;
      }
      return placeValue;
    }
    case "StoreLocal": {
      const placeValue = read(constants, value.value);
      if (placeValue !== null) {
        constants.set(value.lvalue.place.identifier.id, placeValue);
      }
      return placeValue;
    }
    case "FunctionExpression": {
      constantPropagationImpl(value.loweredFunc.func, constants);
      return null;
    }
    default: {
      // TODO: handle more cases
      return null;
    }
  }
}

/**
 * Recursively read the value of a place: if it is a constant place, attempt to read
 * from that place until reaching a primitive or finding a value that is unset.
 */
function read(constants: Constants, place: Place): Constant | null {
  return constants.get(place.identifier.id) ?? null;
}

type Constant = Primitive | LoadGlobal;
type Constants = Map<IdentifierId, Constant>;

File: compiler/crates/react_optimization/src/constant_propagation.rs
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

use std::collections::HashMap;

use react_diagnostics::Diagnostic;
use react_estree::{BinaryOperator, JsValue};
use react_hir::{
    initialize_hir, merge_consecutive_blocks, BlockKind, Environment, Function, GotoKind,
    GotoTerminal, IdentifierId, InstructionValue, LoadGlobal, Primitive, TerminalValue,
};
use react_ssa::eliminate_redundant_phis;

pub fn constant_propagation(env: &Environment, fun: &mut Function) -> Result<(), Diagnostic> {
    let mut constants = Constants::default();
    constant_propagation_impl(env, fun, &mut constants)
}

fn constant_propagation_impl(
    env: &Environment,
    fun: &mut Function,
    constants: &mut Constants,
) -> Result<(), Diagnostic> {
    loop {
        let have_terminals_changed = apply_constant_propagation(env, fun, constants)?;
        if !have_terminals_changed {
            break;
        }
        // If terminals have changed then blocks may have become newly unreachable,
        // so reinitialize the HIR
        // TODO handle errors
        initialize_hir(&mut fun.body).unwrap();

        // Now that predecessors have changed, prune phi operands for unreachable blocks
        // for example, a phi node whose operand was eliminated because it was set in a
        // block that is no longer reached
        for block in fun.body.blocks.iter_mut() {
            // TODO: avoid the clone here
            let predecessors = block.predecessors.clone();
            for phi in block.phis.iter_mut() {
                phi.operands
                    .retain(|predecessor, _| predecessors.contains(predecessor))
            }
        }

        // By removing some phi operands, there may be phis that were not previously
        // redundant but now are
        eliminate_redundant_phis(env, fun);

        // Finally, merge together any blocks that are now guaranteed to execute
        // consecutively
        merge_consecutive_blocks(env, fun)?;
    }
    Ok(())
}

fn apply_constant_propagation(
    env: &Environment,
    fun: &mut Function,
    constants: &mut Constants,
) -> Result<bool, Diagnostic> {
    let mut has_changes = false;

    for block in fun.body.blocks.iter_mut() {
        for phi in block.phis.iter() {
            let mut value: Option<Constant> = None;
            for (_, operand) in &phi.operands {
                if let Some(operand_value) = constants.get(operand.id) {
                    match &mut value {
                        Some(value) if value == operand_value => {
                            // no-op
                        }
                        Some(_) => {
                            value = None;
                            break;
                        }
                        None => {
                            value = Some(operand_value.clone());
                        }
                    }
                } else {
                    // This phi operand's value is unknown, bail out of replacing it
                    value = None;
                    break;
                }
            }
            if let Some(value) = value {
                constants.insert(phi.identifier.id, value);
            }
        }
        for (ix, instr_ix) in block.instructions.iter().enumerate() {
            if block.kind == BlockKind::Sequence && ix == block.instructions.len() - 1 {
                // Evaluating the last value of a sequence can break order of evaluation
                // so skip these instructions
                continue;
            }
            let instr_ix = usize::from(*instr_ix);
            let lvalue_id = fun.body.instructions[instr_ix].lvalue.identifier.id;
            let mut value = std::mem::replace(
                &mut fun.body.instructions[instr_ix].value,
                InstructionValue::Tombstone,
            );
            let const_value = evaluate_instruction(env, &mut value, constants)?;
            if let Some(const_value) = const_value {
                constants.insert(lvalue_id, const_value);
            }
            fun.body.instructions[instr_ix].value = value;
        }

        // If the block ends in an `if` and the test value is a constant primitive,
        // then convert the terminal into a goto to either the consequent or alternate
        // in this case, only the selected branch is reachable
        if let TerminalValue::If(terminal) = &mut block.terminal.value {
            if let Some(primitive) = constants.get_primitive(terminal.test.identifier.id) {
                let target_block_id = if primitive.value.is_truthy() {
                    terminal.consequent
                } else {
                    terminal.alternate
                };
                block.terminal.value = TerminalValue::Goto(GotoTerminal {
                    block: target_block_id,
                    kind: GotoKind::Break,
                });
                has_changes = true;
            }
        }
    }

    Ok(has_changes)
}

fn evaluate_instruction(
    env: &Environment,
    mut instr: &mut InstructionValue,
    constants: &mut Constants,
) -> Result<Option<Constant>, Diagnostic> {
    match &mut instr {
        InstructionValue::Primitive(value) => Ok(Some(Constant::Primitive(value.clone()))),
        InstructionValue::LoadGlobal(value) => Ok(Some(Constant::Global(value.clone()))),
        InstructionValue::Binary(value) => {
            let left = constants.get_primitive(value.left.identifier.id);
            let right = constants.get_primitive(value.right.identifier.id);
            match (left, right) {
                (Some(left), Some(right)) => {
                    if let Some(result) =
                        apply_binary_operator(env, &left.value, value.operator, &right.value)
                    {
                        *instr = InstructionValue::Primitive(Primitive {
                            value: result.clone(),
                        });
                        Ok(Some(Constant::Primitive(Primitive { value: result })))
                    } else {
                        Ok(None)
                    }
                }
                _ => {
                    // no-op, not all operands are known
                    Ok(None)
                }
            }
        }
        InstructionValue::LoadLocal(value) => {
            if let Some(const_value) = constants.get(value.place.identifier.id) {
                *instr = const_value.into();
                Ok(Some(const_value.clone()))
            } else {
                Ok(None)
            }
        }
        InstructionValue::StoreLocal(value) => {
            if let Some(const_value) = constants.get(value.value.identifier.id).cloned() {
                constants.insert(value.lvalue.identifier.identifier.id, const_value.clone());
                Ok(Some(const_value))
            } else {
                Ok(None)
            }
        }
        InstructionValue::Function(value) => {
            // TODO: due to the outer fixpoint iteration this could visit the same
            // function many times. However we only strictly have to visit the function
            // again if the context variable's constant values have changed since last
            // time. Improve this by tracking the inner_constants value with which we
            // last visited, and skip visiting if the same
            let mut inner_constants: Constants = value
                .lowered_function
                .context
                .iter()
                .filter_map(|id| {
                    let value = constants.get(id.identifier.id);
                    value.map(|value| (id.identifier.id, value.clone()))
                })
                .collect();
            constant_propagation_impl(env, &mut value.lowered_function, &mut inner_constants)?;
            Ok(None)
        }
        _ => {
            // no-op, not all instructions can be processed
            Ok(None)
        }
    }
}

fn apply_binary_operator(
    _env: &Environment,
    left: &JsValue,
    operator: BinaryOperator,
    right: &JsValue,
) -> Option<JsValue> {
    match (left, right) {
        (JsValue::Number(left), JsValue::Number(right)) => match operator {
            BinaryOperator::Add => Some(JsValue::Number(*left + *right)),
            BinaryOperator::Subtract => Some(JsValue::Number(*left - *right)),
            BinaryOperator::Multiply => Some(JsValue::Number(*left * *right)),
            BinaryOperator::Divide => Some(JsValue::Number(*left / *right)),
            BinaryOperator::LessThan => Some(JsValue::Boolean(*left < *right)),
            BinaryOperator::LessThanOrEqual => Some(JsValue::Boolean(*left <= *right)),
            BinaryOperator::GreaterThan => Some(JsValue::Boolean(*left > *right)),
            BinaryOperator::GreaterThanOrEqual => Some(JsValue::Boolean(*left >= *right)),
            BinaryOperator::Equals => Some(JsValue::Boolean(left.equals(*right))),
            BinaryOperator::NotEquals => Some(JsValue::Boolean(left.not_equals(*right))),
            BinaryOperator::StrictEquals => Some(JsValue::Boolean(left.equals(*right))),
            BinaryOperator::NotStrictEquals => Some(JsValue::Boolean(left.not_equals(*right))),
            _ => None,
        },
        (left, right) => match operator {
            BinaryOperator::Equals => left.loosely_equals(right).map(JsValue::Boolean),
            BinaryOperator::NotEquals => left.not_loosely_equals(right).map(JsValue::Boolean),
            BinaryOperator::StrictEquals => Some(JsValue::Boolean(left.strictly_equals(right))),
            BinaryOperator::NotStrictEquals => {
                Some(JsValue::Boolean(left.not_strictly_equals(right)))
            }
            _ => None,
        },
    }
}

#[derive(Default)]
struct Constants {
    data: HashMap<IdentifierId, Constant>,
}

impl Constants {
    fn get_primitive(&self, id: IdentifierId) -> Option<&Primitive> {
        if let Some(Constant::Primitive(primitive)) = &self.data.get(&id) {
            Some(primitive)
        } else {
            None
        }
    }

    fn get(&self, id: IdentifierId) -> Option<&Constant> {
        self.data.get(&id)
    }

    fn insert(&mut self, id: IdentifierId, constant: Constant) {
        self.data.insert(id, constant);
    }
}

impl FromIterator<(IdentifierId, Constant)> for Constants {
    fn from_iter<T: IntoIterator<Item = (IdentifierId, Constant)>>(iter: T) -> Self {
        Self {
            data: FromIterator::from_iter(iter),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
enum Constant {
    Global(LoadGlobal),
    Primitive(Primitive),
}

impl From<&Constant> for InstructionValue {
    fn from(value: &Constant) -> Self {
        match value {
            Constant::Global(value) => InstructionValue::LoadGlobal(value.clone()),
            Constant::Primitive(value) => InstructionValue::Primitive(value.clone()),
        }
    }
}

impl From<Constant> for InstructionValue {
    fn from(value: Constant) -> Self {
        match value {
            Constant::Global(value) => InstructionValue::LoadGlobal(value),
            Constant::Primitive(value) => InstructionValue::Primitive(value),
        }
    }
}
